docs/manual/resource-leak-checker.tex

```diff
@@ -588,6 +597,234 @@
 will only use \<@MustCallUnknown> as a default when encountering a language
 feature that it is unable to reason about.
 
+\sectionAndLabel{Arrays of resources}{resource-leak-arrays}
+To understand how the checker handles arrays, let us first remind ourselves what happens if we assign resources with non-empty must-call type to an array:
+
+\begin{verbatim}
+Socket[] s = new Socket[n];
+s[0] = new Socket(myHost, myPort);
+\end{verbatim}
+
+The assignment throws an error, because the scope of the constructed Socket ends, without its must-call obligations having been fulfilled.
+
+\begin{verbatim}
+Example.java:31: error: [required.method.not.called] @MustCall method close may not have been invoked on new Socket(myHost, myPort) or any of its aliases.
+            arr[0] = new Socket(myHost, myPort);
+                     ^
+  The type of object is: java.net.Socket.
+  Reason for going out of scope: regular method exit
+1 error
+\end{verbatim}
+
+We need a way for the array to take on this obligation for the resource. This functionality is jointly provided by the \<MustCallOnElements> and \<CalledMethodsOnElements> subcheckers.
+
+\subsectionAndLabel{Declaration and assignment of resource-holding arrays}{resource-leak-arrays-declaration}
+Annotating the array with the \<@OwningArray> declaration annotation allows it to take on such obligations by using a pattern-matched assignment syntax:
+
+\begin{verbatim}
+@OwningArray Socket[] s = new Socket[n];
+// 1. @MustCallOnElements({}) @CalledMethodsOnElements({})
+for (int i = 0; i < s.length; i++) {
+  try {
+    s[i] = new Socket(myHost, myPort);
+  } catch (Exception e) {
+    // do nothing
+  }
```

Comment:
If we allow for the `new Socket` expression to fail with an `Exception`, shouldn't we also allow for a `null` check in the loop that calls `close()` on all the elements?  Since if `new Socket` throws an `Exception`, `s[i]` will be `null`

Reply:
yes we could do that. It would be purely syntactic. I'll write it down to do later, let's check on some real code first whether this is usually done. Maybe there's going to be a ton of other checks that are sometimes done, which we will have to support and we should take a more systematic approach than simple syntactic pattern-matching and case distinctions.